pkg/reduce/reduce_test.go

```diff
@@ -143,26 +146,107 @@ func TestDataForward_Start(t *testing.T) {
+
+	var err error
+	var pbqManager *pbq.Manager
+
+	// create pbqManager
+	pbqManager, err = pbq.NewManager(child, pbq.WithPBQStoreOptions(store.WithStoreSize(100), store.WithPbqStoreType(dfv1.InMemoryType)),
+		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
+	assert.NoError(t, err)
+
+	// create new fixed window of (windowTime)
+	window := fixed.NewFixed(windowTime)
+
+	var reduceDataForwarder *DataForward
+	reduceDataForwarder, err = NewDataForward(child, forwardReduceTest{}, fromBuffer, toBuffer, pbqManager, forwardReduceTest{}, f, p, window, WithReadBatchSize(10))
+	assert.NoError(t, err)
+
+	go reduceDataForwarder.Start(child)
+
+	for to.IsEmpty() {
 		select {
-		case <-ticker.C:
-			// build 10 messages from the start time, time difference between the messages is 1 second
-			messages := buildMessagesForReduce(count, key+strconv.Itoa(i))
-			i++
+		case <-child.Done():
+			assert.Fail(t, child.Err().Error())
+		default:
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
 
-			// TODO use watermark in memory store when it is ready to synthesize offsets
-			// write the messages to fromBuffer, so that it will be available for consuming
-			fromBuffer.Write(ctx, messages)
-		case <-ctx.Done():
-			ticker.Stop()
-			return
+	// we are reading only one message here but the count should be equal to
+	// the number of keyed windows that closed
+	msgs, readErr := to.Read(child, 1)
+
+	assert.Nil(t, readErr)
+	assert.Len(t, msgs, 1)
+
+	// assert the output of reduce (count of messages)
+	var readMessagePayload PayloadForTest
+	_ = json.Unmarshal(msgs[0].Payload, &readMessagePayload)
+	// here the count will always be equal to the number of messages written per key
+	assert.Equal(t, 10, readMessagePayload.Value)
+	assert.Equal(t, "count", readMessagePayload.Key)
+
+}
+
+// fetcherAndPublisher creates watermark fetcher and publishers for toBuffers, and keeps the processors alive by sending heartbeats
+func fetcherAndPublisher(toBuffers map[string]isb.BufferWriter) (fetch.Fetcher, map[string]publish.Publisher) {
+
+	var (
+		ctx          = context.Background()
+		keyspace     = "reduce"
+		pipelineName = "testPipeline"
+		hbBucketName = keyspace + "_PROCESSORS"
+		otBucketName = keyspace + "_OT"
+	)
+
+	publishEntity := processor.NewProcessorEntity("reduceProcessor")
+	sourcePublishEntity := processor.NewProcessorEntity("sourceProcessor")
+	hb, hbWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, hbBucketName)
+	ot, otWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, otBucketName)
+
+	publishers := make(map[string]publish.Publisher)
+
+	p := publish.NewPublish(ctx, publishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+	sourcePublisher := publish.NewPublish(ctx, sourcePublishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+
+	go func() {
```

Comment:
why do we need this? can't we put watermark while writing messages?

Reply:
this is for heartbeat